### PR TITLE
VMware cluster states + dependencies

### PR DESCRIPTION
**Pull request by alexbleotu**
### What does this PR do?

Adds 3 states and all required dependencies (+ a missing external VSAN library from VMware) to be used with the `esxcluster` proxy to configure a VMware cluster, rename the cluster's datastore, assign a VSAN license to the cluster

Functions added:

State functions:

- cluster_configured state to create/manage ESX clusters
- vsan_datastore_configured esxcluster state that renames the vSan datastore on a cluster
-  licenses_configured state that checks the the VSAN licenses are added to the vCenter and 

Execution functions

- salt.modules.vspere.list_datastores_via_proxy that returns a list of datastore representations
- salt.modules.vspere.rename_datastore that renames a datastore
- salt.modules.vsphere.list_licenses that lists all licenses on a vCenter
- salt.modules.vsphere.add_license
- salt.modules.vsphere._get_entity that describes a cluster and a vcenter
- salt.modules.vsphere._validate_entity that validates a cluster and a vcenter description
- salt.modules.vsphere.list_assigned_licenses that list the licenses assigned to an entity
- salt.modules.vsphere.assign_license that assigns a license to an entity

Utils functions:

- salt.utils.vmware.get_storage_system that retrieves an ESXi host's storage system
- salt.utils.vmware.get_datastores which returns a list of vim.Datastore objects
- salt.utils.vmware.rename_datastore
- salt.utils.vmware.get_license_manager
- salt.utils.vmware.get_license_assignment_manager
- salt.utils.vmware.get_licenses
- salt.utils.vmware.add_license
- salt.utils.vmware.get_assigned_licenses
- salt.utils.vmware.assign_license

### Tests written?

Yes (utils functions only)

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Review by cachedout — CHANGES_REQUESTED** [review]
Fantastic! Let's start with the pylint fixes and then we'll start reviewing this in-depth. Thanks @alexbleotu!

https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/14781/violations/

File: salt/states/esxcluster.py [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+    pip install pyVmomi
+
+.. note::
+
+    Version 6.0 of pyVmomi has some problems with SSL error handling on certain
+    versions of Python. If using version 6.0 of pyVmomi, Python 2.6,
+    Python 2.7.9, or newer must be present. This is due to an upstream dependency
+    in pyVmomi 6.0 that is not supported in Python versions 2.7 to 2.7.8. If the
+    version of Python is not in the supported range, you will need to install an
+    earlier version of pyVmomi. See `Issue #29537`_ for more information.
```
Comment (cachedout) [inline salt/states/esxcluster.py]: Fair enough but are we detecting this condition and warning the user at all? It's a bit much to ask them to read the docs here to determine what could be a very hard-to-detect condition. I wonder if we could improve this at all.

File: salt/states/esxcluster.py [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+
+
+def vsan_datastore_configured(name, datastore_name):
+    '''
+    Configures the cluster's vsan_datastore
+
+    WARNING: vsan datastores will not exist until there is at least on host in
+    the cluster; the state assumes that the datastore exists and errors out if i
+    it doesn't; it's up to the user to accept the error or enable the state run
+    when de datastore does exist (grain: vsan_datastore_exists)
```
Comment (cachedout) [inline salt/states/esxcluster.py]: `de`? `the` perhaps?

**Review by cachedout — COMMENTED** [review]
This is one of the best PRs I've seen in a long while. Great job, @alexbleotu. Just a few small comments in-line.

File: salt/config/schemas/esxcluster.py [hunk 3]
```diff
@@ -14,9 +14,180 @@
...
+
+    enabled = BooleanItem(
+        title='Enabled',
+        description='Specifies if HA should be enabled')
+    admission_control_enabled = BooleanItem(
+        title='Admission Control Enabled')
+    admission_control_policy = AdmissionControlPolicyItem()
+    default_vm_settings = DefaultVmSettingsItem()
+    hb_ds_candidate_policy = StringItem(
+        title='Hartbeat Datastore Candidate Policy',
```
[the remark this answers is not in the document]
Comment (alexbleotu) [reply salt/config/schemas/esxcluster.py]: Good catch

[the remark this answers is not in the document]
Comment (alexbleotu) [reply salt/config/schemas/esxcluster.py]: Good catch!

File: salt/states/esxcluster.py (hunk 1 above)
Comment (alexbleotu) [reply salt/states/esxcluster.py]: I am not sure how to do this though. The only thing I can think of is warnings in tests if there is an incompatibility, but there is no guarantee that everyone will run the tests.

File: salt/states/esxcluster.py (hunk 2 above)
Comment (alexbleotu) [reply salt/states/esxcluster.py]: There are several typos; will fix ;)

File: salt/states/esxcluster.py (hunk 1 above)
Comment (rallytime) [reply salt/states/esxcluster.py]: Can we detect this in `__virtual__()` and `return False, 'State module did not load: Incompatible versions of Python and pyVmomi present'`, or something along those lines?

Also, this is minor, but I think the mention of Python 2.6 can be removed to reduce potential confusion. Support for Python 2.6 in salt has been removed as of the 2017.7.0 release.

**Comment by alexbleotu** [comment]
https://github.com/saltstack/salt/pull/43674 depends on this

File: salt/states/esxcluster.py (hunk 1 above)
Comment (alexbleotu) [reply salt/states/esxcluster.py]: We should be able to. This needs to happen in all states though and the `vsphere` execution module. I will add it in this one to `esxcluster` and I will create a separate PR to fix the rest

Comment (alexbleotu) [reply salt/states/esxcluster.py]: Ok, I have a change - not as trivial as expected. Testing it out now. Question is if we should add it in the state files or just in the `vsphere` execution module where pyVmomi is actually used. @rallytime, @cachedout what do you think?

Comment (alexbleotu) [reply salt/states/esxcluster.py]: Ok did both. Let's see if the tests pass. From what I remember they don't actually test the `__virtual__` function

**Review by cachedout — APPROVED** [review]

**Comment by cachedout** [comment]
@rallytime Re-review please

**Review by rallytime — APPROVED** [review]
Well done! And thank you for adding tests, too. :)

**Pull request merged**